server/controllers/azuread.js

```diff
@@ -0,0 +1,134 @@
+  `https://login.microsoftonline.com/${tenantId}/oauth2/v2.0/token`;
+const OAUTH_USER_INFO_ENDPOINT = "https://graph.microsoft.com/oidc/userinfo";
+const OAUTH_GRANT_TYPE = "authorization_code";
+const OAUTH_RESPONSE_TYPE = "code";
+
+async function azureAdSignIn(ctx) {
+  const config = configValidation();
+  const redirectUri = encodeURIComponent(config["AZUREAD_OAUTH_REDIRECT_URI"]);
+  const endpoint = OAUTH_ENDPOINT(config["AZUREAD_TENANT_ID"]);
+
+  // Generate code verifier and code challenge
+  const { code_verifier: codeVerifier, code_challenge: codeChallenge } =
+    pkceChallenge();
+
+  // Store the code verifier in the session
+  ctx.session.codeVerifier = codeVerifier;
+
+  const url = `${endpoint}?client_id=${config["AZUREAD_OAUTH_CLIENT_ID"]}&redirect_uri=${redirectUri}&scope=${config["AZUREAD_SCOPE"]}&response_type=${OAUTH_RESPONSE_TYPE}&code_challenge=${codeChallenge}&code_challenge_method=S256`;
+  ctx.set("Location", url);
+  return ctx.send({}, 302);
+}
+
+async function azureAdSignInCallback(ctx) {
+  const config = configValidation();
+  const tokenService = getService("token");
+  const userService = getService("user");
+  const oauthService = strapi.plugin("strapi-plugin-sso").service("oauth");
+  const roleService = strapi.plugin("strapi-plugin-sso").service("role");
+
+  if (!ctx.query.code) {
+    return ctx.send(oauthService.renderSignUpError(`code Not Found`));
+  }
+
+  const params = new URLSearchParams();
+  params.append("code", ctx.query.code);
+  params.append("client_id", config["AZUREAD_OAUTH_CLIENT_ID"]);
+  params.append("client_secret", config["AZUREAD_OAUTH_CLIENT_SECRET"]);
+  params.append("redirect_uri", config["AZUREAD_OAUTH_REDIRECT_URI"]);
+  params.append("grant_type", OAUTH_GRANT_TYPE);
+
+  // Include the code verifier from the session
+  params.append("code_verifier", ctx.session.codeVerifier);
+
+  try {
+    const tokenEndpoint = OAUTH_TOKEN_ENDPOINT(config["AZUREAD_TENANT_ID"]);
+    const response = await axios.post(tokenEndpoint, params, {
+      headers: {
+        "Content-Type": "application/x-www-form-urlencoded",
+      },
+    });
+    const userResponse = await axios.get(OAUTH_USER_INFO_ENDPOINT, {
+      headers: {
+        Authorization: `Bearer ${response.data.access_token}`,
+      },
+    });
+
+    const dbUser = await userService.findOneByEmail(userResponse.data.email);
+    let activateUser;
+    let jwtToken;
+
+    if (dbUser) {
+      activateUser = dbUser;
+      jwtToken = await tokenService.createJwtToken(dbUser);
+    } else {
+      const azureAdRoles = await roleService.azureAdRoles();
+      const roles =
+        azureAdRoles && azureAdRoles["roles"]
+          ? azureAdRoles["roles"].map((role) => ({
+              id: role,
+            }))
+          : [];
+
+      const defaultLocale = oauthService.localeFindByHeader(
+        ctx.request.headers
+      );
+      activateUser = await oauthService.createUser(
+        userResponse.data.email,
```

Comment:
I don't seem to be able to get the email in my environment.  Is this a problem with my Azure settings? Or do I need to modify the code?

Reply:
I believe you have to enable the email, given_name, family_name optional claims.

https://learn.microsoft.com/en-us/azure/active-directory/develop/active-directory-optional-claims#configuring-optional-claims

Reply:
Thank you.
It seems that this one was not retrieved by the API because the email address was not set for the logged in user.